src/fp.rs

```diff
@@ -158,6 +245,110 @@ impl<'a, 'b> Mul<&'b Fp> for &'a Fp {
 impl_binops_additive!(Fp, Fp);
 impl_binops_multiplicative!(Fp, Fp);
 
+impl<T> core::iter::Sum<T> for Fp
+where
+    T: core::borrow::Borrow<Fp>,
+{
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = T>,
+    {
+        iter.fold(Self::zero(), |acc, item| acc + item.borrow())
+    }
+}
+
+impl<T> core::iter::Product<T> for Fp
+where
+    T: core::borrow::Borrow<Fp>,
+{
+    fn product<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = T>,
+    {
+        iter.fold(Self::one(), |acc, item| acc * item.borrow())
+    }
+}
+
+impl Field for Fp {
+    const ZERO: Self = Fp::zero();
+    const ONE: Self = Fp::one();
+
+    fn random(rng: impl RngCore) -> Self {
+        Self::random(rng)
+    }
+
+    fn square(&self) -> Self {
+        self.square()
+    }
+
+    fn double(&self) -> Self {
+        self.add(self)
+    }
+
+    fn invert(&self) -> CtOption<Self> {
+        self.invert()
+    }
+
+    fn sqrt_ratio(num: &Self, div: &Self) -> (Choice, Self) {
+        ff::helpers::sqrt_ratio_generic(num, div)
+    }
+
+    fn sqrt(&self) -> CtOption<Self> {
+        self.sqrt()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ReprFp(pub(crate) [u8; 48]);
+
+impl AsMut<[u8]> for ReprFp {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0[..]
+    }
+}
+
+impl AsRef<[u8]> for ReprFp {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+impl Default for ReprFp {
+    fn default() -> Self {
+        Self([0u8; 48])
+    }
+}
+
+impl PrimeField for Fp {
+    type Repr = ReprFp;
+
+    fn from_repr(r: Self::Repr) -> CtOption<Self> {
+        Self::from_bytes(&r.0)
```

Comment:
I think `from_bytes` and `to_bytes` are using big-endian, so we might need to port the `from_repr` and `to_repr` impl here as well

Reply:
I think we'd need to perform a montgomery reduction to get canonical representation. Perhaps we could copy-paste from `from_bytes` and `to_bytes`:

```rust
impl PrimeField for Fp {
    type Repr = ReprFp;

    fn from_repr(repr: Self::Repr) -> CtOption<Self> {
        let mut tmp = Fp([0, 0, 0, 0, 0, 0]);

        tmp.0[0] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[0..8]).unwrap());
        tmp.0[1] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[8..16]).unwrap());
        tmp.0[2] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[16..24]).unwrap());
        tmp.0[3] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[24..32]).unwrap());
        tmp.0[4] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[32..40]).unwrap());
        tmp.0[5] = u64::from_le_bytes(<[u8; 8]>::try_from(&repr.0[40..48]).unwrap());

        // Try to subtract the modulus
        let (_, borrow) = sbb(tmp.0[0], MODULUS[0], 0);
        let (_, borrow) = sbb(tmp.0[1], MODULUS[1], borrow);
        let (_, borrow) = sbb(tmp.0[2], MODULUS[2], borrow);
        let (_, borrow) = sbb(tmp.0[3], MODULUS[3], borrow);
        let (_, borrow) = sbb(tmp.0[4], MODULUS[4], borrow);
        let (_, borrow) = sbb(tmp.0[5], MODULUS[5], borrow);

        // If the element is smaller than MODULUS then the
        // subtraction will underflow, producing a borrow value
        // of 0xffff...ffff. Otherwise, it'll be zero.
        let is_some = (borrow as u8) & 1;

        // Convert to Montgomery form by computing
        // (a.R^0 * R^2) / R = a.R
        tmp *= &R2;

        CtOption::new(tmp, Choice::from(is_some))
    }

    fn to_repr(&self) -> Self::Repr {
        // Turn into canonical form by computing
        // (a.R) / R = a
        let tmp = Fp::montgomery_reduce(
            self.0[0], self.0[1], self.0[2], self.0[3], self.0[4], self.0[5], 0, 0, 0, 0, 0, 0,
        );

        let mut res = [0; 48];
        res[0..8].copy_from_slice(&tmp.0[0].to_le_bytes());
        res[8..16].copy_from_slice(&tmp.0[1].to_le_bytes());
        res[16..24].copy_from_slice(&tmp.0[2].to_le_bytes());
        res[24..32].copy_from_slice(&tmp.0[3].to_le_bytes());
        res[32..40].copy_from_slice(&tmp.0[4].to_le_bytes());
        res[40..48].copy_from_slice(&tmp.0[5].to_le_bytes());

        ReprFp(res)
    }

    // ...
}
```

And I'd suggest to add a test for `from_repr` and `to_repr` like:

```rust
#[test]
fn test_repr() {
    for value in [Fp::ZERO, Fp::ONE, Fp::ROOT_OF_UNITY] {
        assert_eq!(Fp::from_repr(value.to_repr()).unwrap(), value);
    }
}
```

Also I think the wrapper `ReprFp` seems a little bit verbose to me so I'd propose to just have `[u8; 48]` as the `Repr`, which is what other fields are using.

Reply:
Unfortunately `[u8;48]` does not implement `Default` bound. So this was also the case with extension field representatoins

Reply:
Ah I see it now

Reply:
So what I've done is basically add LE for `ReprFp` and BE for the `from/to_bytes`. I'll try to document it somewhere and that should be it. We should have the two flavours.

Is this ok?